spec.html

```diff
@@ -4421,7 +4417,28 @@ <h1>Runtime Semantics: MV</h1>
               The MV of <emu-grammar>StrUnsignedDecimalLiteral ::: DecimalDigits ExponentPart</emu-grammar> is the MV of |DecimalDigits| times 10<sup>_e_</sup>, where _e_ is the MV of |ExponentPart|.
             </li>
           </ul>
-          <p>Once the exact MV for a String numeric literal has been determined, it is then rounded to a value of the Number type. If the MV is 0, then the rounded value is *+0*<sub>𝔽</sub> unless the first non white space code point in the String numeric literal is `-`, in which case the rounded value is *-0*<sub>𝔽</sub>. Otherwise, the rounded value must be the Number value for the MV (in the sense defined in <emu-xref href="#sec-ecmascript-language-types-number-type"></emu-xref>), unless the literal includes a |StrUnsignedDecimalLiteral| and the literal has more than 20 significant digits, in which case the Number value may be either the Number value for the MV of a literal produced by replacing each significant digit after the 20th with a 0 digit or the Number value for the MV of a literal produced by replacing each significant digit after the 20th with a 0 digit and then incrementing the literal at the 20th digit position. A digit is significant if it is not part of an |ExponentPart| and</p>
+          <p>For the MV of any production not shown above, see <emu-xref href="#sec-static-semantics-mv"></emu-xref>.</p>
+        </emu-clause>
+
+        <emu-clause id="sec-stringtonumber" aoid="StringToNumber">
+          <h1>StringToNumber ( _str_ )</h1>
+          <p>The abstract operation StringToNumber takes argument _str_ (a String) and returns a Number. It performs the following steps when called:</p>
+          <emu-alg>
+            1. Let _text_ be ! StringToCodePoints(_str_).
+            1. Let _literal_ be ParseText(_text_, |StringNumericLiteral|).
+            1. If _literal_ is a List of errors, return *NaN*.
+            1. If _literal_ contains a |StrUnsignedDecimalLiteral| and _literal_ has more than 20 significant digits, then
```

Comment:
Is this supposed to be `Contains`?

```suggestion
            1. If _literal_ Contains |StrUnsignedDecimalLiteral| and _literal_ has more than 20 significant digits, then
```

Also, would it be better to refer to the number of `DecimalDigit`s in `_literal_` since it's a parse node?

Reply:
> Is this supposed to be `Contains`?

Nope. At least, I certainly didn't intend `Contains`. In my opinion, `Contains` is a weird thing that shouldn't be used unless you actually need its weirdness. The ordinariness of its name is misleading. You should maybe think of it as `ContainsIfYouIgnoreCertainSubtrees`.

The corresponding point in the status quo actually says "includes", but it's rare for the spec to use "includes" in this way (I think there's only one other such example), and more common to use "contains". 
 
> Also, would it be better to refer to the number of `DecimalDigit`s in `_literal_` since it's a parse node?

You mean replace "significant digits" with "significant |DecimalDigit|s"? I'm not sure if that would be better.

Reply:
Okay, I'm fine with the usage of "contains" in this way. We should look into that other usage of "includes" that you mention. It might be nice to normalise that phrasing.

The significant digits part still bothers me a bit. There's a type mismatch. How about this?

> ... and the source text matched by |StrUnsignedDecimalLiteral| has more than 20 significant digits, then

Also, how many significant digits does `Infinity` have? Maybe we should exclude that case in the statement somehow?

Reply:
> We should look into that other usage of "includes" that you mention. It might be nice to normalise that phrasing.

(It's in [28.3 Module Namespace Objects](https://tc39.es/ecma262/#sec-module-namespace-objects): "an _ImportDeclaration_ that includes a _NameSpaceImport_".)

> The significant digits part still bothers me a bit. There's a type mismatch. How about this?
> 
> > ... and the source text matched by |StrUnsignedDecimalLiteral| has more than 20 significant digits, then

Yeah, I think that'd be okay.

> Also, how many significant digits does `Infinity` have? Maybe we should exclude that case in the statement somehow?

Well, it doesn't have any digits in the first place, so it can't have any significant digits.

I agree that the whole "rounding to 20 significant digits" thing is somewhat loose (relying on the reader's intuition more than the spec normally does), but tightening it up isn't the point of this PR. If a future PR can factor out the two occurrences of rounding-to-20, then that would probably be a better time to look at how it's defined.

Reply:
> I agree that the whole "rounding to 20 significant digits" thing is somewhat loose

How so? It gives a definition which looks pretty precise to me:

> A digit is significant if it is not part of an |ExponentPart| and
> - it is not `0`; or
> - there is a non-zero digit to its left and there is a non-zero digit, not in the `ExponentPart`, to its right.

(And indeed I think the only way to interpret this is that `Infinity` has no digits and so has no significant digits.)

Reply:
Yeah, the definition of significant digit is okay, although it would be a bit more precise if it related "digit" to "DecimalDigit" (as @michaelficarra suggested). I'm more thinking about the context in which it's used. E.g. "incrementing the literal at the 20th digit position" relies on the reader's intuition rather that precisely specifying the effect in terms of source text.

Reply:
I'm fine with not making the Infinity exception because, while not obvious, I agree that it is well-defined as zero.

> "incrementing the literal at the 20th digit position" relies on the reader's intuition rather that precisely specifying the effect in terms of source text

I agree. We could probably just modify it to increment a numeric interpretation of the source text somehow if we want to be very pedantic here.

Reply:
Note that I'm not arguing for or against the looseness, I'm just saying (a) it's there, and (b) deciding whether to do something about it should be left out of this PR. (So I agree with @bakkot that this is an aside.)